src/transformers/models/llama/modeling_llama.py

```diff
@@ -96,6 +97,21 @@ def _expand_mask(mask: torch.Tensor, dtype: torch.dtype, tgt_len: Optional[int]
     return inverted_mask.masked_fill(inverted_mask.to(torch.bool), torch.finfo(dtype).min)
 
 
+def _inplace_unmask_padding(expanded_mask, attention_mask):
+    bsz = attention_mask.shape[0]
+
+    # Get the index of the first non-zero value for every sample in the batch.
+    tmp = torch.arange(attention_mask.shape[1], 0, -1)
+    indices = torch.argmax(attention_mask.cpu() * tmp, 1, keepdim=True)
+
+    max_len = torch.max(indices)
+    range_tensor = torch.arange(max_len).unsqueeze(0)
+    range_tensor = range_tensor.repeat(indices.size(0), 1)
+    range_tensor[range_tensor >= indices] = 0
+
+    expanded_mask[torch.arange(bsz).unsqueeze(1), 0, range_tensor] = 0
```

Comment:
would be nice to have a comment saying which places are unmasked. Somehow hard to read through the block to get to know it.

(ok, I see you mentioned it unmasks all places - but having comments would ease the life 🙏 )

Reply:
@ydshieh I'd be curious to have your opinion on the current design. Currently the thing is we addition masks in `_prepare_decoder_attention_mask`, namely:

```python
            combined_attention_mask = (
                expanded_attn_mask if combined_attention_mask is None else expanded_attn_mask + combined_attention_mask
            )
```

The addition of `_inplace_unmask_padding` thus sets part of the lines on the axis -2 to `0.0` but the rightmost part of these lines are still masked by the causal mask from `_make_causal_mask`. Thus the current implementation may be "breaking" in the sense that the output logits for the padding tokens may not be identical as before.

A cleaner solution would be, for padding tokens on dimension -2, to unmask all the tokens (ignore causality there). With this solution, we may be able to get the same logits as the softmax would be computed on `[0., 0., 0., 0. ]` instead of on `[-3.4e+38, -3.4e+38, -3.4e+38, -3.4e+38]`.

This would however require to rewrite `_make_causal_mask`, `_expand_mask` which would not behave as before - which I would say is fine given that those functions are private.

WDYT?

Reply:
If we want to keep this thing in a standard (that we define customly), the best is to unmask after we have computed all the masks. This means, after computing `combined_attention_mask = expanded_attn_mask + combined_attention_mask`, we unmask `combined_attention_mask`, by using the information of both `attention_mask` and causal mask (in the `[0,1]` format).

However, as we compute with `-3.4e+38` too early (IMO), and we no longer have this `[0,1]` information (well, we have this for `attention_mask`).

**Maybe we can simply check if a line has being all masked by checking if all entries for that line has `value <= dtype.finfo.min`. And for those line, unmask it (i.e. just before `combined_attention_mask`)**

The problem of such modification(s) (to keep things explicit defined) is that it introduces overhead, and I am sure other reviewers / people may not appreciate this overhead. **So I suggest having opinion from core maintainers (and patrick too) before more changes.**

(Rewrite `_make_causal_mask` and `_expand_mask` could work, but better not to do it.)

Reply:
Thank you, it sounds sensible. This would basically apply `_inplace_unmask_padding` & `combined_attention_mask[torch.arange(bsz).unsqueeze(1), 0, range_tensor] = 0` after `combined_attention_mask = expanded_attn_mask + combined_attention_mask` instead of previously as I currently did.

We would indeed need to be careful about the overhead, but given the vectorized implementation (no for loop) it may be fine. Currently this is guarded by `input_shape[-1] > 1 and is_torch_sdpa_available() and attention_mask.device.type == "cuda"`.

Reply:
Should be fixed!